pkg/cache/sql/informer/listoption_indexer.go

```diff
@@ -352,14 +439,30 @@ func (l *ListOptionIndexer) ListByOptions(ctx context.Context, lo ListOptions, p
 		offsetClause = "\n  OFFSET ?"
 		params = append(params, offset)
 	}
+	if limit == 0 && offset == 0 {
+		countQuery = ""
+		countParams = []any{}
+	}
```

Comment:
In the current implementation, if `limit == 0 && offset == 0` we don't run a count statement.

Also in the current implementation, constructing the SQL query and running it are done in one function,
so they're interleaved, which makes query generation more cumbersome to test -- you need to test it
through mocks, rather than just looking at the string a function call returns.

I wanted to move all the query construction into one function, so the best way to tell the executor not to
run the count query is to set it to the empty string.

So I maintain the new code isn't changing the database lookups.

Reply:
> In the current implementation, if limit == 0 && offset == 0 we don't run a count statement.

Agreed

> Also in the current implementation, constructing the SQL query and running it are done in one function,
so they're interleaved, which makes query generation more cumbersome to test -- you need to test it
through mocks, rather than just looking at the string a function call returns.

Agreed

> I wanted to move all the query construction into one function

Agreed, it's cleaner now

> So I maintain the new code isn't changing the database lookups.

Agreed

> However, I noticed that there's no need to include the ORDER clause in a SELECT COUNT stmt, so I dropped it.

Agreed

> the best way to tell the executor not to run the count query is to set it to the empty string

Meh... I personally prefer not to mess with mutation unless necessary, so I'd rather have all struct members be written once. Also, an argument could be made that it's the job of the executor to execute - and so potentially to decide when not to.

Anyway, I understand taste and style play a role here. Up to you to make the final call to either change this or just resolve this conversation.

Reply:
I looked at this, but it's simpler to create `countQuery` based on the first query, which refers only to tables and fields and doesn't have any of the pagination stuff.

Let's try building the final `QueryInfo` block a bit at a time, which should give us both more simplicity and clarity.

In fact I see I'm setting `queryInfo := &QueryInfo{}` at the start of `constructQuery` and then reassigning it. 
 `golangci-lint` doesn't flag that error, and does flag a lot of things that the `steve .golangci.json` says to ignore, so I don't think we need to add it to lasso right now.

Changing this required a few changes to the unit tests, mainly distinguishing undefined slices as nil vs an empty slice.